iavl_nodedb.go

```diff
@@ -73,68 +107,418 @@ func (ndb *nodeDB) SaveNode(node *IAVLNode) {
+			k := make([]byte, len(it.Key()))
+			v := make([]byte, len(it.Value()))
+
+			// Leveldb reuses the memory, we are forced to copy.
+			copy(k, it.Key())
+			copy(v, it.Value())
+
+			fn(k, v)
+		}
+		if err := it.Error(); err != nil {
+			cmn.PanicSanity(err.Error())
+		}
+		it.Release()
+	} else {
+		ndb.traverse(func(key, value []byte) {
+			if bytes.HasPrefix(key, prefix) {
+				fn(key, value)
+			}
+		})
 	}
+}
 
-	if elem, ok := ndb.cache[string(node.hash)]; ok {
-		ndb.cacheQueue.Remove(elem)
-		delete(ndb.cache, string(node.hash))
+func (ndb *nodeDB) uncacheNode(hash []byte) {
+	if elem, ok := ndb.nodeCache[string(hash)]; ok {
+		ndb.nodeCacheQueue.Remove(elem)
+		delete(ndb.nodeCache, string(hash))
 	}
-	ndb.orphans[string(node.hash)] = struct{}{}
 }
 
 // Add a node to the cache and pop the least recently used node if we've
 // reached the cache size limit.
 func (ndb *nodeDB) cacheNode(node *IAVLNode) {
-	elem := ndb.cacheQueue.PushBack(node)
-	ndb.cache[string(node.hash)] = elem
+	elem := ndb.nodeCacheQueue.PushBack(node)
+	ndb.nodeCache[string(node.hash)] = elem
 
-	if ndb.cacheQueue.Len() > ndb.cacheSize {
-		oldest := ndb.cacheQueue.Front()
-		hash := ndb.cacheQueue.Remove(oldest).(*IAVLNode).hash
-		delete(ndb.cache, string(hash))
+	if ndb.nodeCacheQueue.Len() > ndb.nodeCacheSize {
+		oldest := ndb.nodeCacheQueue.Front()
+		hash := ndb.nodeCacheQueue.Remove(oldest).(*IAVLNode).hash
+		delete(ndb.nodeCache, string(hash))
 	}
 }
 
-// Write to disk. Orphans are deleted here.
+// Write to disk.
 func (ndb *nodeDB) Commit() {
 	ndb.mtx.Lock()
 	defer ndb.mtx.Unlock()
 
-	// Delete orphans from previous block
-	for orphanHashStr, _ := range ndb.orphansPrev {
-		ndb.batch.Delete([]byte(orphanHashStr))
-	}
-	// Write saves & orphan deletes
 	ndb.batch.Write()
-	ndb.db.SetSync(nil, nil)
 	ndb.batch = ndb.db.NewBatch()
-	// Shift orphans
-	ndb.orphansPrev = ndb.orphans
-	ndb.orphans = make(map[string]struct{})
+}
+
+func (ndb *nodeDB) getRoots() ([][]byte, error) {
+	roots := [][]byte{}
+
+	ndb.traversePrefix([]byte(rootsPrefix), func(k, v []byte) {
+		roots = append(roots, v)
+	})
+	return roots, nil
+}
+
+// SaveRoot creates an entry on disk for the given root, so that it can be
+// loaded later.
+func (ndb *nodeDB) SaveRoot(root *IAVLNode, version uint64) error {
+	ndb.mtx.Lock()
+	defer ndb.mtx.Unlock()
+
+	if len(root.hash) == 0 {
+		cmn.PanicSanity("Hash should not be empty")
+	}
+	if version <= ndb.getLatestVersion() {
+		return errors.New("can't save root with lower or equal version than latest")
+	}
+
+	// Note that we don't use the version attribute of the root. This is
+	// because we might be saving an old root at a new version in the case
+	// where the tree wasn't modified between versions.
+	key := fmt.Sprintf(rootsPrefixFmt, version)
+	ndb.batch.Set([]byte(key), root.hash)
+	ndb.cacheVersion(version, root.hash)
+
+	return nil
+}
+
+///////////////////////////////////////////////////////////////////////////////
```

Comment:
Let's note that these functions are only used for testing. 

If any are used outside testing, we should move them, or make it clear.

Reply:
Done. 60d9992